sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,759 @@
+  mmio_region_t base_addr;
+  dif_usbdev_buffer_pool_t buffer_pool;
+  dif_usbdev_read_buffer_op_state_t read_buffer_op_state;
+  /**
+   * Active buffer operations.
+   *
+   * Bits 0..11: True if writing bytes of an outgoing packet for endpoint N
+   * (0..11).
+   * Bits 12..30: Unused.
+   * Bit 31: True if reading bytes of an incoming packet.
+   */
+  uint32_t active_buffer_ops;
+} dif_usbdev_t;
+
+/**
+ * Enumeration for enabling/disabling various functionality.
+ */
+typedef enum dif_usbdev_toggle {
+  kDifUsbdevToggleDisable,
+  kDifUsbdevToggleEnable,
+} dif_usbdev_toggle_t;
+
+/**
+ * Set of allowed configurations for USB power sense override.
+ */
+typedef enum dif_usbdev_power_sense_override {
+  kDifUsbdevPowerSenseOverrideDisabled,
+  kDifUsbdevPowerSenseOverridePresent,
+  kDifUsbdevPowerSenseOverrideNotPresent
+} dif_usbdev_power_sense_override_t;
+
+/**
+ * Configuration for initializing a USB device.
+ */
+typedef struct dif_usbdev_config {
+  mmio_region_t base_addr;
+  dif_usbdev_toggle_t differential_rx;
+  dif_usbdev_toggle_t differential_tx;
+  dif_usbdev_toggle_t single_bit_eop;
+  dif_usbdev_power_sense_override_t power_sense_override;
+} dif_usbdev_config_t;
+
+/**
+ * Common return codes for the functions in this library.
+ *
+ * `kDifUsbdevResultInvalidArgument` signals that there was a problem with the
+ * argument(s) but the function did not cause any side effects.
+ * `kDifUsbdevResultGeneralError` covers other error cases.
+ */
+typedef enum dif_usbdev_result {
+  kDifUsbdevResultOK,
+  kDifUsbdevResultInvalidArgument,
+  kDifUsbdevResultGeneralError,
+} dif_usbdev_result_t;
+
+/**
+ * Initialize a USB device.
+ *
+ * A USB device must first be initialized by this function before calling other
+ * functions in this library.
+ *
+ * @param config Configuration for initializing a USB device.
+ * @param usbdev Internal state of the initialized USB device.
+ * @return `kDifUsbdevResultInvalidArgument` if `config` or `usbdev` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `config` is invalid, `kDifUsbdevResultOK`
+ * otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_init(dif_usbdev_config_t *config,
+                                    dif_usbdev_t *usbdev);
+
+/**
+ * Fill the available buffer FIFO of a USB device.
+ *
+ * The USB device has a small FIFO (AV FIFO) that stores free buffers for
+ * incoming packets. It is the responsibility of the software to ensure that the
+ * AV FIFO is never empty. If the host tries to send a packet when the AV FIFO
+ * is empty, the USB device will respond with a NAK. While this will typically
+ * cause the host to retry transmission for regular data packets, there are
+ * transactions in the USB protocol during which the USB device is not allowed
+ * to send a NAK. Thus, the software must make sure that the AV FIFO is never
+ * empty by calling this function periodically.
+ *
+ * @param usbdev A USB device.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
```

Comment:
Nit: I would tend to suggest something a bit more terse for the return code, personally.

Reply:
I find it simpler to define an `enum` with error codes, comment all of the variants, and in the function description then simply do:
`@return dif_<ip>_result_t.`